### ISSUE TITLE
Negative notified_serial crashes pdns_server and pdnsutil check-all-zones

### ISSUE THREAD

**Issue by opeter** [PowerDNS/pdns#4475]
Somehow a negative notified_serial sneaked into our domains table.  It causes pdns_server to crash/loop, should be easy to reproduce and easy to fix:

In our case:

```
mysql> select notified_serial from domains where notified_serial is not null order by notified_serial asc limit 10;  
+-----------------+
| notified_serial |
+-----------------+
|       -93865895 |
|               4 |
|      2006013100 |
|      2006050100 |
|      2006061900 |
|      2006061900 |
|      2006090500 |
|      2006090500 |
|      2006090500 |
|      2006091800 |
+-----------------+
10 rows in set (1.04 sec)

mysql> select \* from domains where notified_serial = -93865895;
+--------+------------+---------------+------------+--------+-----------------+---------+
| id     | name       | master        | last_check | type   | notified_serial | account |
+--------+------------+---------------+------------+--------+-----------------+---------+
| 845349 | xxxxxxx.de | xxx.xxx.xxx.x |       NULL | MASTER |       -93865895 | NULL    |
+--------+------------+---------------+------------+--------+-----------------+---------+
1 row in set (1.71 sec)

mysql> update domains set notified_serial = NULL where id = 845349 limit 1;  
Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```

The Error messages are:

```
    Sep 15 06:45:07 a pdns[14120]: PowerDNS Authoritative Server 4.0.1 (C) 2001-2016 PowerDNS.COM BV                                                                                                                              
    Sep 15 06:45:07 a pdns[14120]: Using 64-bits mode. Built using clang 3.4.1 (tags/RELEASE_34/dot1-final 208032) on Sep 15 2016 08:35:02 by root@localhost                                                                      
    Sep 15 06:45:07 a pdns[14120]: PowerDNS comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it according to the terms of the GPL version 2.                                         
    Sep 15 06:45:07 a pdns[14120]: Creating backend connection for TCP                                                                                                                                                            
    Sep 15 06:45:07 a pdns[14120]: Master/slave communicator launching                                                                                                                                                            
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    Sep 15 06:45:07 a pdns[14120]: About to create 8 backend threads for UDP                                                                                                                                                      
    [...]                                                                                                                                                                                                                         
    Sep 15 06:45:09 a pdns[14120]: Exiting because communicator thread died with STL error: stou                                                                                                                                  
    Sep 15 06:45:09 a kernel: pid 14120 (pdns_server), uid 1001: exited on signal 8                                                                                                                                               
    Sep 15 06:45:09 a pdns[93674]: Our pdns instance (14120) exited after signal 8                                                                                                                                                
    Sep 15 06:45:09 a pdns[93674]: Respawning  
```

And:

```
    # pdnsutil check-all-zones                                                                                                                                                                                 
    Error: stou                                                                                                                                                                                                                   
    # pdnsutil -v check-all-zones                                                                                                                                                                              
    Error: stou  
```

**Comment by cmouse** [comment]
Related to #4450

**Comment by webtic** [comment]
Not only negative numbers give this error; I had some zones with a serial in the format of YYYYMMDDHHMMSS doing the same. Although these numbers obviously overflow 4294967296 they did work on earlier versions of PowerDNS.

**Comment by zeha** [comment]
They would "work" by overflowing the serial field in the SOA in outgoing packets, though.

**Comment by webtic** [comment]
The silent overflowing is indeed the reason I did not notice it earlier. I added this case to this thread as after upgrading to 4.0 suddenly PowerDNS got stuck in a restart loop due to the fact that this stou error appeared on zones which were "ok" for older versions.

So as an headsup for others puzzled by the same error: not only negative but also overflowing serials can cause this error.

**Comment by opeter** [comment]
What about using the database (in my case MySQL) to do the sanatizing by an abs() envelope?

Any non-positive number or string will be returned as 0 and overwritten with the current timestamp by the daemon.

```
diff --git a/modules/gmysqlbackend/gmysqlbackend.cc b/modules/gmysqlbackend/gmysqlbackend.cc
index 1be4e1e..6562852 100644
--- a/modules/gmysqlbackend/gmysqlbackend.cc
+++ b/modules/gmysqlbackend/gmysqlbackend.cc
@@ -92,9 +92,9 @@ public:
 
     declare(suffix,"master-zone-query","Data", "select master from domains where name=? and type='SLAVE'");
 
-    declare(suffix,"info-zone-query","","select id,name,master,last_check,notified_serial,type,account from domains where name=?");
+    declare(suffix,"info-zone-query","","select id,name,master,abs(last_check) as last_check,abs(notified_serial) as notified_serial,type,account from domains where name=?");
 
-    declare(suffix,"info-all-slaves-query","","select id,name,master,last_check from domains where type='SLAVE'");
+    declare(suffix,"info-all-slaves-query","","select id,name,master,abs(last_check) as last_check from domains where type='SLAVE'");
     declare(suffix,"supermaster-query","", "select account from supermasters where ip=? and nameserver=?");
     declare(suffix,"supermaster-name-to-ips", "", "select ip,account from supermasters where nameserver=? and account=?");
 
@@ -119,7 +119,7 @@ public:
     declare(suffix,"update-serial-query","", "update domains set notified_serial=? where id=?");
     declare(suffix,"update-lastcheck-query","", "update domains set last_check=? where id=?");
     declare(suffix,"zone-lastchange-query", "", "select max(change_date) from records where domain_id=?");
-    declare(suffix,"info-all-master-query","", "select id,name,master,last_check,notified_serial,type from domains where type='MASTER'");
+    declare(suffix,"info-all-master-query","", "select id,name,master,abs(last_check) as last_check,abs(notified_serial) as notified_serial,type from domains where type='MASTER'");
     declare(suffix,"delete-domain-query","", "delete from domains where name=?");
     declare(suffix,"delete-zone-query","", "delete from records where domain_id=?");
     declare(suffix,"delete-rrset-query","","delete from records where domain_id=? and name=? and type=?");
@@ -142,7 +142,7 @@ public:
     declare(suffix,"delete-tsig-key-query","", "delete from tsigkeys where name=?");
     declare(suffix,"get-tsig-keys-query","", "select name,algorithm, secret from tsigkeys");
 
-    declare(suffix, "get-all-domains-query", "Retrieve all domains", "select domains.id, domains.name, records.content, domains.type, domains.master, domains.notified_serial, domains.last_check, domains.account from domains LEFT JOIN records ON records.domain_id=domains.id AND records.type='SOA' AND records.name=domains.name WHERE records.disabled=0 OR ?");
+    declare(suffix, "get-all-domains-query", "Retrieve all domains", "select domains.id, domains.name, records.content, domains.type, domains.master, abs(domains.notified_serial) as domains.notified_serial, abs(domains.last_check) as domains.last_check, domains.account from domains LEFT JOIN records ON records.domain_id=domains.id AND records.type='SOA' AND records.name=domains.name WHERE records.disabled=0 OR ?");
 
     declare(suffix, "list-comments-query", "", "SELECT domain_id,name,type,modified_at,account,comment FROM comments WHERE domain_id=?");
     declare(suffix, "insert-comment-query", "", "INSERT INTO comments (domain_id, name, type, modified_at, account, comment) VALUES (?, ?, ?, ?, ?, ?)");
```

**Comment by Habbie** [comment]
Putting () around the numbers in records.content SOA also triggers the stou crash.

**Comment by Habbie** [comment]
#5915 is related

**Comment by Habbie** [comment]
On current master, `rectify-all-zones` aborts with `Error: stoul returned result out of unsigned int range; data was "-5"`

**Comment by Habbie** [comment]
The pdns_server crash was fixed in #7460
